src/main/java/org/htmlunit/javascript/host/URLSearchParams.java

```diff
@@ -137,26 +143,146 @@ public URLSearchParams() {
      */
     @JsxConstructor
     public URLSearchParams(final Object params) {
-        // TODO: Pass in a sequence
-        // new URLSearchParams([["foo", 1],["bar", 2]]);
-
-        // TODO: Pass in a record
-        // new URLSearchParams({"foo" : 1 , "bar" : 2});
-
         url_ = new URL("http://www.htmlunit.org", "");
 
         if (params == null || Undefined.isUndefined(params)) {
             return;
         }
 
         try {
-            url_.setSearch(splitQuery(Context.toString(params)));
+            url_.setSearch(resolveParams(params));
+        }
+        catch (final EcmaError e) {
+            throw ScriptRuntime.typeError("Failed to construct 'URLSearchParams': " + e.getErrorMessage());
         }
         catch (final MalformedURLException e) {
             LOG.error(e.getMessage(), e);
         }
     }
 
+    private static List<NameValuePair> resolveParams(final Object params) {
+        // if params is a sequence
+        if (params instanceof Scriptable && ScriptableObject.hasProperty((Scriptable)params, SymbolKey.ITERATOR)) {
+            final List<NameValuePair> query = new ArrayList<>();
+
+            final Scriptable sequence = (Scriptable)params;
+            final Iterator<Object> sequenceIterator = new IterationProtocolIterator(sequence, Context.getCurrentContext(), sequence.getParentScope());
+            while (sequenceIterator.hasNext()) {
+                final Object innerSequence = sequenceIterator.next();
```

Comment:
how about we reduce the number of concepts? :shrug:
```suggestion
            final Iterator<Object> paramsIterator = new IterationProtocolIterator((Scriptable)params,
                    Context.getCurrentContext(), ((Scriptable)params).getParentScope());
            while (sequenceIterator.hasNext()) {
                final Object nameValue = paramsIterator.next();
```
this makes the outer code and inner code a bit more consistent